### PR TITLE
FEAT: single session/login

### PR DESCRIPTION
**Pull request by siinghd**
This pr should add support #101 

Key changes need to be added: 

ENV: set  `JWT_SECRET`, correctly as its being used

DB: user model has new field called `token`
`NEXT_PUBLIC_BASE_URL_LOCAL="http://localhost:3000"` -> set this to the correct one

Middleware: check for the token mismatch to redirect the user to signout. 

Currently with nextauth you cannot logout the user on server side so im just redirecting the user to a page and doing singout in the useeffect.

Also middleware runtime is edge so it doesnt  support db querying directly from there  unless you don't use prisma accelerate and prisma edge client.

Missing: admin can logout the user

### CODE REVIEW

**Comment by siinghd** [comment]
wait before mergin, let me add also the admin part and test it properly

**Comment by siinghd** [comment]
ok, should be working fine, 

tested at https://incomparable-biscotti-b0e8a5.netlify.app

@hkirat NEXT_PUBLIC_BASE_URL_LOCAL(should be https://app.100xdevs.com , i think), JWT_SECRET needs to be set, NEXTAUTH_URL needs to be set too correct url (https://app.100xdevs.com) i think.


https://github.com/code100x/cms/assets/38940603/e831d99f-8186-4d66-a76f-b70962e06876

admin:
Screenshot 2024-02-26 at 12 29 14 PM

File: src/app/invalidsession/page.tsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
+'use client';
+
+import { signOut } from 'next-auth/react';
+import React, { useEffect } from 'react';
+
+const page = () => {
+  useEffect(() => {
+    signOut({
+      callbackUrl: '/signin',
+    });
```
Comment (devsargam) [inline src/app/invalidsession/page.tsx:10]:
```suggestion
    });
    toast('Too many devices connected. Logging out!', {
      action: {
        label: 'Close',
        onClick: () => console.log('Closed Toast'),
      },
    });
```

File: src/lib/auth.ts [hunk 2]
```diff
@@ -1,6 +1,20 @@
 import db from '@/db';
 import CredentialsProvider from 'next-auth/providers/credentials';
+import { SignJWT, importJWK } from 'jose';
```
Comment (devsargam) [inline src/lib/auth.ts:4]:
```suggestion
import { Toaster } from '@/components/ui/sonner';

```

File: src/app/invalidsession/page.tsx [hunk 3]
```diff
@@ -0,0 +1,16 @@
+'use client';
+
+import { signOut } from 'next-auth/react';
+import React, { useEffect } from 'react';
+
```
Comment (devsargam) [inline src/app/invalidsession/page.tsx:5]:
```suggestion
import { Toaster } from '@/components/ui/sonner';

```

File: src/app/invalidsession/page.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+const page = () => {
+  useEffect(() => {
+    signOut({
+      callbackUrl: '/signin',
+    });
+  }, []);
+
+  return <div>page</div>;
```
Comment (devsargam) [inline src/app/invalidsession/page.tsx:13]:
```suggestion
  return <div><Toaster /></div>;
```

**Comment by devsargam** [comment]
@siinghd displaying a toast to the end user would be a nice UX. Could you please accept those changes

File: src/middleware.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  matcher: ['/courses/:path*'],
+};
+
+export default withAuth(async (req) => {
+  const token = req.nextauth.token;
+  if (!token) {
+    return NextResponse.redirect(new URL('/invalidsession', req.url));
+  }
+  const user = await fetch(
+    `${process.env.NEXT_PUBLIC_BASE_URL_LOCAL}/api/user?token=${token.jwtToken}`,
```
[the remark this answers is not in the document]
Comment (siinghd) [reply src/middleware.ts:14]: to use prisma in middleware we need to use accelerate and prisma edge, https://github.com/prisma/prisma/issues/21310, middleware runs on edge runtime

**Comment by hkirat** [comment]
/bounty $150
Thankyou sir. One more comment but merging it in for now

**Pull request merged**